cmd/tmpfs/tmpfs.go

```diff
@@ -0,0 +1,347 @@
+	"net"
+	"os"
+	"sync"
+
+	"github.com/Harvey-OS/go/internal/tmpfs"
+	"github.com/Harvey-OS/ninep/pkg/debugfs"
+	"github.com/Harvey-OS/ninep/protocol"
+)
+
+var (
+	networktype = flag.String("ntype", "tcp4", "Default network type")
+	netaddr     = flag.String("addr", ":5641", "Network address")
+	debug       = flag.Int("debug", 0, "Print debug messages")
+	pkg         = flag.String("package", "", "tar.gz package to open")
+)
+
+type fileServer struct {
+	Versioned bool
+
+	archive    *tmpfs.Archive
+	filesMutex sync.Mutex
+	files      map[protocol.FID]*FidEntry
+	ioUnit     protocol.MaxSize
+}
+
+// FidEntry wraps an Entry with the instance data required for a fid reference
+type FidEntry struct {
+	tmpfs.Entry
+
+	// Username to be used for all entries in this hierarchy
+	uname string
+
+	// We can't know how big a serialized dentry is until we serialize it.
+	// At that point it might be too big. We save it here if that happens,
+	// and on the next directory read we start with that.
+	oflow []byte
+
+	// Index of next child to return when reading a directory
+	nextChildIdx int
+}
+
+func newFidEntry(entry tmpfs.Entry, uname string) *FidEntry {
+	return &FidEntry{Entry: entry, uname: uname, oflow: nil, nextChildIdx: -1}
+}
+
+func (fs *fileServer) Rversion(msize protocol.MaxSize, version string) (protocol.MaxSize, string, error) {
+	if version != "9P2000" {
+		return 0, "", fmt.Errorf("%v not supported; only 9P2000", version)
+	}
+	return msize, version, nil
+}
+
+// TODO handle aname?
+func (fs *fileServer) Rattach(fid protocol.FID, afid protocol.FID, uname string, aname string) (protocol.QID, error) {
+	if afid != protocol.NOFID {
+		return protocol.QID{}, fmt.Errorf("We don't do auth attach")
+	}
+
+	// There should be no .. or other such junk in the Aname. Clean it up anyway.
+	/*aname = path.Join("/", aname)
+	st, err := os.Stat(aname)
+	if err != nil {
+		return protocol.QID{}, err
+	}*/
+
+	root := fs.archive.Root()
+	fs.filesMutex.Lock()
+	fs.files[fid] = newFidEntry(root, uname)
+	fs.filesMutex.Unlock()
+
+	return root.Qid(), nil
+}
+
+func (fs *fileServer) Rflush(o protocol.Tag) error {
+	return nil
+}
+
+func (fs *fileServer) Rwalk(fid protocol.FID, newfid protocol.FID, paths []string) ([]protocol.QID, error) {
+	// Lookup the parent fid
+	parentEntry, err := fs.getFile(fid)
+	if err != nil {
+		return nil, fmt.Errorf("does not exist")
+	}
+
+	if len(paths) == 0 {
+		// Clone fid - point to same entry
+		fs.filesMutex.Lock()
+		defer fs.filesMutex.Unlock()
+		_, ok := fs.files[newfid]
+		if ok {
+			return nil, fmt.Errorf("FID in use: clone walk, fid %d newfid %d", fid, newfid)
+		}
+
+		fs.files[newfid] = newFidEntry(parentEntry.Entry, parentEntry.uname)
+		return []protocol.QID{}, nil
+	}
+
+	walkQids := make([]protocol.QID, len(paths))
+
+	var i int
+	var pathcmp string
+	currEntry := parentEntry.Entry
+	for i, pathcmp = range paths {
+		var ok bool
+		var dir *tmpfs.Directory
+		if dir, ok = currEntry.(*tmpfs.Directory); ok {
```

Comment:
you don't use them outside the if; why not
if dir, ok :- etc. etc. and ditch the two var statements above?